lib/std/priority_dequeue.zig

```diff
@@ -0,0 +1,879 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const sort = std.sort;
+const assert = std.debug.assert;
+const warn = std.debug.warn;
+const testing = std.testing;
+const expect = testing.expect;
+const expectEqual = testing.expectEqual;
+const expectError = testing.expectError;
+
+/// Priority Dequeue for storing generic data. Initialize with `init`.
+pub fn PriorityDequeue(comptime T: type) type {
+    return struct {
+        const Self = @This();
+
+        items: []T,
+        len: usize,
+        allocator: *Allocator,
+        lessThanFn: fn (a: T, b: T) bool,
+
+        /// Initialize and return a new dequeue. Provide `lessThanFn`
+        /// that returns `true` when its first argument should
+        /// get min-popped before its second argument. For example,
+        /// to make `popMin` return the minimum value, provide
+        ///
+        /// `fn lessThanFn(a: T, b: T) bool { return a < b; }`
+        pub fn init(allocator: *Allocator, lessThanFn: fn (T, T) bool) Self {
+            return Self{
+                .items = &[_]T{},
+                .len = 0,
+                .allocator = allocator,
+                .lessThanFn = lessThanFn,
+            };
+        }
+
+        fn lessThan(self: Self, a: T, b: T) bool {
+            return self.lessThanFn(a, b);
+        }
+
+        fn greaterThan(self: Self, a: T, b: T) bool {
+            return self.lessThanFn(b, a);
+        }
+
+        /// Free memory used by the dequeue.
+        pub fn deinit(self: Self) void {
+            self.allocator.free(self.items);
+        }
+
+        /// Insert a new element, maintaining priority.
+        pub fn add(self: *Self, elem: T) !void {
+            try ensureCapacity(self, self.len + 1);
+            addUnchecked(self, elem);
+        }
+
+        /// Add each element in `items` to the dequeue.
+        pub fn addSlice(self: *Self, items: []const T) !void {
+            try self.ensureCapacity(self.len + items.len);
+            for (items) |e| {
+                self.addUnchecked(e);
+            }
+        }
+
+        fn addUnchecked(self: *Self, elem: T) void {
+            self.items[self.len] = elem;
+
+            if (self.len > 0) {
+                const start = self.getStartForSiftUp(elem, self.len);
+                self.siftUp(start);
+            }
+
+            self.len += 1;
+        }
+
+        fn isMinLayer(index: usize) bool {
+            // In the min-max heap structure:
+            // The first element is on a min layer;
+            // next two are on a max layer;
+            // next four are on a min layer, and so on.
+            const leading_zeros = @clz(usize, index + 1);
+            const highest_set_bit = 63 - leading_zeros;
```

Comment:
`63` should be `@bitSizeOf(usize) - 1` ?

Reply:
Yes, thank you. I meant to find the right function for this.